src/Configuration/Dotenv/Parse.hs

```diff
@@ -16,43 +16,65 @@
 
 module Configuration.Dotenv.Parse (configParser) where
 
+import Configuration.Dotenv.ParsedVariable
 import Control.Applicative
 import Control.Monad
 import Text.Megaparsec
 import Text.Megaparsec.String (Parser)
 import qualified Text.Megaparsec.Lexer as L
 
+data QuoteType = SingleQuote | DoubleQuote
+
 -- | Returns a parser for a Dotenv configuration file. Accepts key and value
 -- arguments separated by @=@. Comments in all positions are handled
 -- appropriately.
-configParser :: Parser [(String, String)]
+configParser :: Parser [ParsedVariable]
 configParser = between scn eof (sepEndBy1 envLine (eol <* scn))
 
 -- | Parse a single environment variable assignment.
-envLine :: Parser (String, String)
-envLine = (,) <$> (lexeme variableName <* lexeme (char '=')) <*> lexeme value
+envLine :: Parser ParsedVariable
+envLine = ParsedVariable <$> (lexeme variableName <* lexeme (char '=')) <*> lexeme value
 
 -- | Variables must start with a letter or underscore, and may contain
 -- letters, digits or '_' character after the first character.
-variableName :: Parser String
+variableName :: Parser VName
 variableName = ((:) <$> firstChar <*> many otherChar) <?> "variable name"
   where
     firstChar = char '_'  <|> letterChar
     otherChar = firstChar <|> digitChar
 
 -- | Value: quoted or unquoted.
-value :: Parser String
+value :: Parser VValue
 value = (quotedValue <|> unquotedValue) <?> "variable value"
   where
-    quotedValue   = quotedWith '\'' <|> quotedWith '\"'
-    unquotedValue = many (noneOf "\'\" \t\n\r")
+    quotedValue   = quotedWith SingleQuote <|> quotedWith DoubleQuote
+    unquotedValue = Unquoted <$> (many $ fragment "\'\" \t\n\r")
```

Comment:
This happens because you should never nest one `many` into another one. Since `many` parses *zero* or more occurences, it succeeds even when its argument parser doesn't match at at all. And if something succeds without consuming input/advancing parser in `many`, it'll run it indefinitely. First `many` is where you say, and the second (inner) `many` is in `literalValueFragment`. I propose the following technique:

* Use `some` intead of many, so fragments can't be empty.
* Empty values just have zero fragments: `[]`.

There are other minor issues, but they are not important at this point, we will talk about them when the PR is more complete. BTW, good work for a person new to Haskell!